components/brave_wallet/browser/tx_state_manager.cc

```diff
@@ -107,41 +109,68 @@ bool TxStateManager::ValueToTxMeta(const base::Value::Dict& value,
   return true;
 }
 
-TxStateManager::TxStateManager(PrefService* prefs)
-    : prefs_(prefs), weak_factory_(this) {}
+TxStateManager::TxStateManager(PrefService* prefs,
+                               value_store::ValueStoreFrontend* store)
+    : prefs_(prefs), initialized_(false), store_(store), weak_factory_(this) {
+  Initialize();
+}
 
 TxStateManager::~TxStateManager() = default;
 
-void TxStateManager::AddOrUpdateTx(const TxMeta& meta) {
-  ScopedDictPrefUpdate update(prefs_, kBraveWalletTransactions);
-  base::Value::Dict& dict = update.Get();
+void TxStateManager::Initialize() {
+  store_->Get(kStorageTransactions, base::BindOnce(&TxStateManager::OnTxsRead,
+                                                   weak_factory_.GetWeakPtr()));
+}
+
+void TxStateManager::OnTxsRead(absl::optional<base::Value> txs) {
+  if (txs) {
+    txs_ = std::move(txs->GetDict());
+  }
+  initialized_ = true;
+  for (auto& observer : observers_) {
+    observer.OnInitialized();
+  }
+}
+
+bool TxStateManager::ScheduleWrite() {
+  if (!initialized_) {
+    VLOG(1) << __FUNCTION__ << "storage is not initialized yet";
+    return false;
+  }
+  store_->Set(kStorageTransactions, base::Value(txs_.Clone()));
+  return true;
+}
+
+bool TxStateManager::AddOrUpdateTx(const TxMeta& meta) {
   const std::string path =
       base::JoinString({GetTxPrefPathPrefix(meta.chain_id()), meta.id()}, ".");
 
-  bool is_add = dict.FindByDottedPath(path) == nullptr;
-  dict.SetByDottedPath(path, meta.ToValue());
+  bool is_add = txs_.FindByDottedPath(path) == nullptr;
+  txs_.SetByDottedPath(path, meta.ToValue());
+  if (!ScheduleWrite()) {
+    return false;
+  }
```

Comment:
It actually changes `txs_` and then fails `ScheduleWrite`. Is that intended?
Can we check `if (!initialized_)` at top each call and do `DCHECK(initialized_)` in `ScheduleWrite` ?

Reply:
I was doing `if (!initialized_)` first and then shorten it like this. I will revert it back.

Reply:
fixed and squash in 04384e3642fa64ea5ba66b930d0a001924c2fbc5